src/cpp/session/SessionSourceDatabase.cpp

```diff
@@ -69,6 +69,10 @@ const char * const SourceDocument::SourceDocumentTypeCpp       = "cpp";
 
 namespace {
 
+// cached mapping of document id to document path (facilitates efficient path
+// lookup)
+std::map<std::string, std::string> s_idToPath;
```

Comment:
I think we have a similar mechanism in `SessionCodeSearch` for mapping ids to paths; see https://github.com/rstudio/rstudio/blob/master/src/cpp/session/modules/SessionCodeSearch.cpp#L968. I wonder if we can reuse code? (I think it's better for this to live in `SessionSourceDatabase` so perhaps it could be exposed from here